src/object/objecttype.c

```diff
@@ -613,9 +615,15 @@ void createNewObjects(object_t t, cte_t *parent,
     /* Create the objects. */
     nextFreeArea = regionBase;
     for (i = 0; i < destLength; i++) {
+        pptr_t new_base = (pptr_t)nextFreeArea + (i << objectSize);
         /* Create the object. */
         /** AUXUPD: "(True, typ_region_bytes (ptr_val \<acute> nextFreeArea + ((\<acute> i) << unat (\<acute> objectSize))) (unat (\<acute> objectSize)))" */
-        cap_t cap = createObject(t, (void *)((word_t)nextFreeArea + (i << objectSize)), userSize, deviceMemory);
+#if defined(__CHERI_PURE_CAPABILITY__)
+        cap_t cap = createObject(t, (void *)cheri_derive_data_cap(regionBase, (ptraddr_t)new_base, totalObjectSize, -1),
+                                    userSize, deviceMemory);
+#else
+        cap_t cap = createObject(t, (void *)new_base, userSize, deviceMemory);
```

Comment:
I think this change will perturb the C correspondence proof of createNewObjects, not in a very interesting way, but will probably need someone from verification to go in there and tune it.

Reply:
The other name assignments look ok otherwise, so if code paths end up the same then the proofs might work once the above is addressed (it's fine as a refactoring, but if you really want the current proofs to pass on this PR then you'll have to `#ifdef` it out).

Reply:
@Xaphiosis That should be fine, I pushed a change to hide that change under the purecap #ifdef and kept it as it is for non-CHERI.